File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>
using namespace std;

vector<string> solution(vector<string> players, vector<string> callings) {
    vector<string> answer;
    unordered_map<string, int> map1;
    for (int i = 0; i < players.size(); i++) {
        map1[players[i]] = i;
    }
    for (int i = 0; i < callings.size(); i++) {
        int tempIdx = map1[callings[i]];
        swap(players[tempIdx], players[tempIdx-1]);
        map1[callings[i]]--;
        map1[players[tempIdx]]++;
    }
    answer = players;
    return answer;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.HashMap;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        HashMap<Integer, String> ranking = new HashMap<>();
        HashMap<String, Integer> record = new HashMap<>();
        for(int i=0;i<players.length;i++){
            ranking.put(i,players[i]);
            record.put(players[i],i);
        }
        
        for(String player : callings){
            int oldRanking = record.get(player);
            int newRanking = oldRanking - 1;
            String fallPlayer = ranking.get(newRanking);
            
            ranking.replace(newRanking,player);
            ranking.replace(oldRanking,fallPlayer);
            
            record.replace(player, newRanking);
            record.replace(fallPlayer, oldRanking);
        }
        
        String[] result = new String[ranking.size()];
        for(int i=0;i<ranking.size();i++){
            result[i] = ranking.get(i);
        }
        
        return result;
    }
}